crates/common/upload/src/upload.rs

```diff
@@ -0,0 +1,348 @@
+use backoff::ExponentialBackoff;
+use camino::Utf8Path;
+use camino::Utf8PathBuf;
+use log::warn;
+use reqwest::header::CONTENT_LENGTH;
+use reqwest::header::CONTENT_TYPE;
+use reqwest::Body;
+use std::time::Duration;
+use tokio::fs::File;
+use tokio_util::codec::BytesCodec;
+use tokio_util::codec::FramedRead;
+
+use crate::error::ErrContext;
+use crate::error::UploadError;
+
+fn default_backoff() -> ExponentialBackoff {
+    // Default retry is an exponential retry with a limit of 5 minutes total.
+    // Let's set some more reasonable retry policy so we don't block the uploads for too long.
+    ExponentialBackoff {
+        initial_interval: Duration::from_secs(15),
+        max_elapsed_time: Some(Duration::from_secs(300)),
+        randomization_factor: 0.1,
+        ..Default::default()
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct UploadInfo {
+    pub url: String,
+    pub auth: Option<Auth>,
+}
+
+impl From<&str> for UploadInfo {
+    fn from(url: &str) -> Self {
+        Self::new(url)
+    }
+}
+
+impl UploadInfo {
+    pub fn new(url: &str) -> Self {
+        Self {
+            url: url.into(),
+            auth: None,
+        }
+    }
+
+    pub fn with_auth(self, auth: Auth) -> Self {
+        Self {
+            auth: Some(auth),
+            ..self
+        }
+    }
+
+    pub fn url(&self) -> &str {
+        self.url.as_str()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Auth {
+    Bearer(String),
+}
+
+#[derive(Debug)]
+pub struct Uploader {
+    source_filename: Utf8PathBuf,
+    backoff: ExponentialBackoff,
+}
+
+impl Uploader {
+    pub fn new(target_path: Utf8PathBuf) -> Self {
+        Self {
+            source_filename: target_path,
+            backoff: default_backoff(),
+        }
+    }
+
+    pub fn set_backoff(&mut self, backoff: ExponentialBackoff) {
+        self.backoff = backoff;
+    }
+
+    pub async fn upload(&self, url: &UploadInfo) -> Result<(), UploadError> {
+        self.upload_request(url).await?;
+
+        Ok(())
+    }
+
+    async fn upload_request(&self, url: &UploadInfo) -> Result<reqwest::Response, UploadError> {
+        let operation = || async {
+            let file = File::open(&self.source_filename)
+                .await
+                .context(format!("Can't open a file {:?}", &self.source_filename))
+                .map_err(backoff::Error::Permanent)?;
+
+            let file_length = file
+                .metadata()
+                .await
+                .context(format!(
+                    "Can't read a file {:?} metadata",
+                    &self.source_filename
+                ))
+                .map_err(backoff::Error::Permanent)?
+                .len();
+
+            let file_body = Body::wrap_stream(FramedRead::new(file, BytesCodec::new()));
+
+            let mut client = reqwest::Client::new()
+                .put(url.url())
+                .header(CONTENT_TYPE, "application/octet-stream")
```

Comment:
question: Do we always want to set it to `application/octet-stream` though? I realize that it's a safe default and then it's the problem of who received that data to figure out what MIME is it, but maybe we can use a crate for detecting MIME type and then use that.
But it's not a priority and could be somewhat complex, so if you agree, then it could be done in another PR some other time.

Reply:
Good point. @Ruadhri17 could you please confirm that the C8y UI is able to properly render the uploaded files as text files in the UI? Else we have a problem and might need to implement some "file content type detection" before setting this.

Reply:
As there is plan to introduce partial PUT, we can also add MIME type detection then. But I agree: it would be much better to  provide type of data we are sending to the server.

Reply:
> Good point. @Ruadhri17 could you please confirm that the C8y UI is able to properly render the uploaded files as text files in the UI? Else we have a problem and might need to implement some "file content type detection" before setting this.

Sure I will test it and if there will be no issues I will request for a review.

Reply:
@Ruadhri17 confirmed that there are no issues with rendering uploaded files in Cumulocity UI.